storage/cache/subtree_cache.go

```diff
@@ -170,3 +192,53 @@ func (s *SubtreeCache) Flush(setSubtree SetSubtreeFunc) error {
 	}
 	return nil
 }
+
+// makeSuffix creates a suffix key for indexing into the subtree's Leaves and
+// InternalNodes maps.
+func makeSuffix(depth int, index *big.Int) ([]byte, error) {
+	// TODO(al): only supports 8 bit subtree sizes currently
+	sfx := make([]byte, 2)
+	if depth >= 8 {
+		return nil, fmt.Errorf("found depth of %d, but we only support depths of 8 currently", depth)
+	}
+	sfx[0] = byte(depth)
+	if index.BitLen() > 0 {
+		sfx[1] = index.Bytes()[0]
+	}
+	return sfx, nil
+}
+
+// populateSubtreeNodes re-creates the subtree's InternalNodes from the subtree
+// Leaves map.
+func (s *SubtreeCache) populateSubtreeNodes(st *storage.SubtreeProto) error {
+	st.InternalNodes = make(map[string][]byte)
+	rootID := storage.NewNodeIDFromHash(st.Prefix)
+	fullTreeDepth := s.treeHasher.Size() * 8
+	leaves := make([]merkle.HStar2LeafHash, 0, len(st.Leaves))
```

Comment:
Rats, yes; I've made it so you can plug in different functions to reconstitute appropriately and make funcs for Logs & Maps.